### ISSUE TITLE
Test Plan for early Feb (2019) release

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode-go#2288]
If you got a notification from this issue, then you are part of the next update to the Go extension.
Thanks and Welcome!

Here, I am listing some of the features and the bug fixes that will go in the next update and for which I need some help in testing. Since I dont work with Go in my day job, I always feel that someone who does Go more than me can help in testing tremendously.

Also mentioned are the folks who logged the bugs/feature requests, those who upvoted it, and those who fixed it by sending PRs as these are the people with the most context on what the issue is.

And when I say "test", you can do a minimum pass of just installing the update and then do nothing special other than going about using VS Code like you do everyday. Or you can test the feature you requested/implemented or the bug you logged/fixed. Or you can test all of the below. Every bit helps.

Drop in a comment in this issue after 2 or 3 days of [installing the update](https://github.com/Microsoft/vscode-go/wiki/Use-the-beta-version-of-the-latest-Go-extension) i.e the beta version with your findings.

### Debugging improvements

* @xiphon, @renannprado, @muravjov, @primalmotion
    * Fix the Pause command. Fixes [Bug 978](https://github.com/Microsoft/vscode-go/issues/978) with [PR 2126](https://github.com/Microsoft/vscode-go/pull/2126)

* @xiphon, @yinfxs, @ericpitcher , @bihe
    * Allow adding/removing breakpoints when debuggee is running. Fixes [Bug 2002](https://github.com/Microsoft/vscode-go/issues/2002) with [PR 2128](https://github.com/Microsoft/vscode-go/pull/2128)

* @jhendrixMSFT, @YoskeOkamoto2, @Vidydy, @christian-boks
    * Display nested content in variables pane. Fixes [Bug 1010](https://github.com/Microsoft/vscode-go/issues/1010) with [PR 2198](https://github.com/Microsoft/vscode-go/pull/2198)

* @jhendrixMSFT, @christian-boks, 
    * Display shadowed variables in variables pane. Fixes [Bug 1974](https://github.com/Microsoft/vscode-go/issues/1974) with [PR 2254](https://github.com/Microsoft/vscode-go/pull/2254)

* @joegilley, @anphung, @tcagan, @mikecowgill, @stffabi, @seanstrickland-wf, @egorse 
   * Fix the slowness during debugging that got introduced a few releases ago, by caching the package info used to call `ListPackageVars` command in delve. [PR 2289](https://github.com/Microsoft/vscode-go/pull/2289)

* @Ashiroq, @vladbarosan, @willfaught
    * New command `Go: Debug Test at Cursor` to debug the test function under the cursor. This provides the same feature as the debug codelens, but in the form of a command. [Feature Request 1088](https://github.com/Microsoft/vscode-go/issues/1088) implemented with [PR 2059](https://github.com/Microsoft/vscode-go/pull/2059)
* @hellozee
    * Use the new respository link for delve. [PR 2277](https://github.com/Microsoft/vscode-go/pull/2277)

* @aggressivepixels
    * Skip installing delve on 32 bit platforms that dont support it. Fixes [Bug 2191](https://github.com/Microsoft/vscode-go/issues/2191) with [PR 2195](https://github.com/Microsoft/vscode-go/pull/2195)

### Others

* @karim
    * When no folder is opened in VS Code, `Go to definition` feature now works on individual files. Fixes [Bug 2246](https://github.com/Microsoft/vscode-go/issues/2246) with [commit 58817b8](https://github.com/Microsoft/vscode-go/commit/58817b85b1609c3d939f9f6b5429817fafe93c69)

* @kidlj 
    * When the main module is opened in VS Code, `Go to definition` feature now works for sub modules as well. Fixes [Bug 2180](https://github.com/Microsoft/vscode-go/issues/2180) with [PR 2262](https://github.com/Microsoft/vscode-go/pull/2262)

* @wbrefvem
    * Run the on save features only for current file. This will improve performance when multiple files are being saved at once, for eg: after find replace across files. Fixes [Bug 2202](https://github.com/Microsoft/vscode-go/issues/2202) with [commit cf0a61c](https://github.com/Microsoft/vscode-go/commit/cf0a61c20b6b883e57c01e344ec943024cbccad7)

*  @katakonst, @jpap 
    * Show test coverage even if the test fails. [Feature Request 2193](https://github.com/Microsoft/vscode-go/issues/2193) implemented with [PR 2263](https://github.com/Microsoft/vscode-go/pull/2263)

*  @kagof, @dhowden
    * Suggest exported member name only at the start of the comment. Fixes [Bug 2063](https://github.com/Microsoft/vscode-go/issues/2063) with [PR 2070](https://github.com/Microsoft/vscode-go/pull/2070)

* @karthikraobr
    * Avoid invalidating code coverage when updating single line comments. [PR 1996](https://github.com/Microsoft/vscode-go/pull/1996)

*  @segevfiner 
    * Use the right documentation in auto-completion widget when there are multiple functions of the same name but on different receivers. Fixes [Bug 2107](https://github.com/Microsoft/vscode-go/issues/2107) with [PR 2215](https://github.com/Microsoft/vscode-go/pull/2215)

* @richardatphilo, @krasoffski 
    * Fix `Find all references` feature when text is selected. Fixes [Bug 2197](https://github.com/Microsoft/vscode-go/issues/2197) with [PR 2226](https://github.com/Microsoft/vscode-go/pull/2226)

**Comment by ramya-rao-a** [comment]
Including other contributors here whose work will be going into the next update, but their work is already tested

@brycekahle, @hellozee, @shreyu86, @oneslash

**Comment by primalmotion** [comment]
I've quickly checked the debugger, it's the first time it works as expected for me :)

**Comment by OneOfOne** [comment]
The debugger seems to work as intended, yay!

**Comment by krasoffski** [comment]
I was not able to reproduce #2197 on my pet projects. With big code base will try a little bit later.

**Comment by ramya-rao-a** [comment]
@kidlj https://github.com/Microsoft/vscode-go/issues/2180 was tracking the bug that Go to definition doesnt work in the files that are part of the sub module. This is fixed.

The case you are trying in your screenshots is the below
- Go to definition on a file that is part of the sub module (this works and this is what #2180 was initially tracking)
- In the file that opens (this will be from the mod cache i.e $GOPATH/pkg/mod), you try Go to definition again. This fails. (Note, this works if initially you were working on the main module and not the sub modules)

I have created #2296 to track this.

**Comment by ramya-rao-a** [comment]
cc @lggomez to take a look at the debugging features and bug fixes, as he has added quite a few features in this area in the past himself.

**Comment by karim** [comment]
@ramya-rao-a I don't know if I should reply here to say everything is working as expected or reply only if there is a problem with the latest beta.

Your fix in #2246 is working great and I tested it using the latest beta. Thanks again. :smiley:

**Comment by ramya-rao-a** [comment]
> I don't know if I should reply here to say everything is working as expected or reply only if there is a problem with the latest beta.

Even feedback that everything is working as expected is very useful. Thanks @karim!

**Comment by lggomez** [comment]
I did a quick rundown test of the debug changes and so far It is good, at least on a minimalist test package. I am on leave until thursday so I wont be able to debug complex and productive tests until then

This release is going to be really great

**Comment by tcagan** [comment]
Hi,
To chime in :

- #2002 - works well - I can now add breakpoint after debug session as started
- #1010 - works well - can see deeply nested structures. Still limited (eventually is stops showing values) but great improvement
- [PR 2289](https://github.com/Microsoft/vscode-go/pull/2289) works well. Stepping through work as expected, no more jumping into proc.go and I don't experience slowness.

Over all - this version is great and debugging experience is much improved - thanks for everyone (@ramya-rao-a specifically) for this!

**Comment by lggomez** [comment]
@tcagan just so you know, recursion depth can be configured in the debugger via settings:

```
"go.delveConfig": {
        "useApiV1": false,
        "dlvLoadConfig": {
            "followPointers": true,
            "maxVariableRecurse": 3,
            "maxStringLen": 100,
            "maxArrayValues": 100,
            "maxStructFields": -1
        }
    }
```

Be careful with the values because the traversal time might grow exponentially, I would not recommend a maxVariableRecurse value higher than 5, but you're welcome to try

**Comment by jhendrixMSFT** [comment]
@tcagan do you have an example where deeply nested structures stops working so I can investigate?  My change should work for structs of arbitrary depth (was the intent anyways).

**Comment by tcagan** [comment]
@jhendrixMSFT - please see attached image. 

I am using [go-jira](https://github.com/andygrunwald/go-jira) and looking at an instance of [Issue](https://github.com/andygrunwald/go-jira/blob/master/issue.go) struct.

My variable name is `issue`.

At the top of the watch pane, you can see it expanded - it shows `issue.ChangeLog.Histories` but does not expand the items within that slice.

Following, I am showing specific item in the `Histories` slice - here it does not expand items in `Items` slice.

Following you can see a specific item in the `Items` slice (just for reference that the items are not empty).

Is this the expected behavior? 
Could it possibly be related to slice/array expansion? 
Maybe I Should change something in the settings?

![screen shot 2019-02-06 at 8 41 23 am](https://user-images.githubusercontent.com/36638770/52324987-08501f00-29ec-11e9-8ce0-50261d1ad6d0.png)

**Comment by tcagan** [comment]
@ramya-rao-a btw, this version also fixed an issue I've had that the debugged process wasn't terminated (for a web api project, I had to kill it manually each time as it was blocking my port and I couldn't re-run)

**Comment by segevfiner** [comment]
#2215 seems working. I wish we could also get the `godef` part working too, but they don't seem responsive to the PR...

**Comment by jhendrixMSFT** [comment]
@tcagan thanks this is helpful.  The problem here is the `eval` expression fails, see the output in the debug console.
```
Failed to evaluate expression - 1:28: expected operand, found 'go'
```
The type name returned by delve, `github.com/andygrunwald/go-jira.ChangelogHistory`, appears to be the problem.  I'll investigate further.

**Comment by ramya-rao-a** [comment]
@jhendrixMSFT Thats a good find, but I believe we should continue investigating that, but dont need to block the build for it. Can you log a separate issue for this and copy over @tcagan's feedback in https://github.com/Microsoft/vscode-go/issues/2288#issuecomment-460919209

@tcagan For https://github.com/Microsoft/vscode-go/issues/2288#issuecomment-460941939, is there a github issue for it?

**Comment by ramya-rao-a** [comment]
The 0.9.0 update is now released, thanks everyone for all the help :)

Happy Coding!

**Comment by dhowden** [comment]
Did this break "run benchmark"?  It seems to be passing the name of benchmark to `go test -run ^X$` instead of doing the old `go test -run ^$ -bench ^X$`

**Comment by ramya-rao-a** [comment]
@dhowden That's right, thanks! Released 0.9.1 with the fix.

**Comment by tcagan** [comment]
@ramya-rao-a I believe its related to #1010 (and fix #2198 does improved it).
Depending on what @jhendrixMSFT finds - it could be related or something new.
